File: analytics/services/groq_provider.py
import os
import httpx


class GroqProvider:
    ENDPOINT = 'https://api.groq.com/openai/v1/chat/completions'
    DEFAULT_MODEL = 'llama-3.3-70b-versatile'
    TIMEOUT = 30.0

    def __init__(self, api_key: str = None, model: str = None):
        self.api_key = api_key or os.environ.get('GROQ_API_KEY', '')
        self.model = model or os.environ.get('GROQ_MODEL', self.DEFAULT_MODEL)
        if not self.api_key:
            raise RuntimeError('GROQ_API_KEY not configured')

    def complete(self, messages: list, options: dict = None) -> str:
        options = options or {}
        body = {
            'model': self.model,
            'messages': messages,
            'temperature': options.get('temperature', 0.1),
            'max_tokens': options.get('max_tokens', 1500),
        }
        if options.get('json_mode'):
            body['response_format'] = {'type': 'json_object'}

        headers = {
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {self.api_key}',
        }

        with httpx.Client(timeout=self.TIMEOUT) as client:
            resp = client.post(self.ENDPOINT, json=body, headers=headers)

        if resp.status_code >= 400:
            try:
                err_msg = resp.json()['error']['message']
            except Exception:
                err_msg = f'HTTP {resp.status_code}'
            raise RuntimeError(f'Groq API error: {err_msg}')

        data = resp.json()
        try:
            return data['choices'][0]['message']['content']
        except (KeyError, IndexError):
            raise RuntimeError('Malformed Groq API response')

    def get_model_name(self) -> str:
        return self.model

File: php_legacy/src/Services/GroqProvider.php
<?php
declare(strict_types=1);

/**
 * GroqProvider
 *
 * Groq uses the OpenAI-compatible chat-completions API, so this implementation
 * would need only a base_url swap to work against OpenAI too.
 *
 * Default model: llama-3.3-70b-versatile
 *   - Strong SQL generation
 *   - Fast inference (Groq's LPU)
 *   - Free tier available
 *
 * Configure via env:
 *   GROQ_API_KEY   (required)
 *   GROQ_MODEL     (optional, default: llama-3.3-70b-versatile)
 */
final class GroqProvider implements LLMProvider
{
    private const ENDPOINT = 'https://api.groq.com/openai/v1/chat/completions';
    private const DEFAULT_MODEL = 'llama-3.3-70b-versatile';
    private const TIMEOUT_SECONDS = 30;

    private string $apiKey;
    private string $model;

    public function __construct(?string $apiKey = null, ?string $model = null)
    {
        $this->apiKey = $apiKey ?? getenv('GROQ_API_KEY') ?: '';
        $this->model  = $model  ?? getenv('GROQ_MODEL') ?: self::DEFAULT_MODEL;

        if ($this->apiKey === '') {
            throw new RuntimeException(
                'GROQ_API_KEY not configured. Set it in .env or config/ai_config.php'
            );
        }
    }

    public function complete(array $messages, array $options = []): string
    {
        $body = [
            'model'       => $this->model,
            'messages'    => $messages,
            'temperature' => $options['temperature'] ?? 0.1,
            'max_tokens'  => $options['max_tokens']  ?? 1500,
        ];

        // Force JSON mode when caller asks for it (Groq supports OpenAI-compatible response_format)
        if (!empty($options['json_mode'])) {
            $body['response_format'] = ['type' => 'json_object'];
        }

        $ch = curl_init(self::ENDPOINT);
        curl_setopt_array($ch, [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_POST           => true,
            CURLOPT_POSTFIELDS     => json_encode($body),
            CURLOPT_TIMEOUT        => self::TIMEOUT_SECONDS,
            CURLOPT_HTTPHEADER     => [
                'Content-Type: application/json',
                'Authorization: Bearer ' . $this->apiKey,
            ],
        ]);

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $curlErr  = curl_error($ch);
        curl_close($ch);

        if ($response === false) {
            throw new RuntimeException("Groq API request failed: {$curlErr}");
        }

        $decoded = json_decode($response, true);

        if ($httpCode >= 400) {
            $msg = "HTTP {$httpCode}";
            if (is_array($decoded) && isset($decoded['error']['message'])) {
                $msg = $decoded['error']['message'];
            } elseif (!empty($response)) {
                $msg .= ': ' . strip_tags(substr($response, 0, 100));
            }
            throw new RuntimeException("Groq API error: {$msg}");
        }

        if (!is_array($decoded) || !isset($decoded['choices'][0]['message']['content'])) {
            throw new RuntimeException('Malformed Groq API response');
        }

        return $decoded['choices'][0]['message']['content'];
    }

    public function getModelName(): string
    {
        return $this->model;
    }
}
